File: modules/cart-price-calc/android/src/main/java/expo/modules/cartpricecalc/CartPriceCalculator.kt
/**
 * Cart Price Calculator — Native port của utils/cart.ts
 * Logic voucher: ALL_REQUIRED, AT_LEAST_ONE_REQUIRED × PERCENT_ORDER, FIXED_VALUE, SAME_PRICE_PRODUCT
 */
package expo.modules.cartpricecalc

import org.json.JSONArray
import org.json.JSONObject

private object VoucherType {
  const val PERCENT_ORDER = "percent_order"
  const val FIXED_VALUE = "fixed_value"
  const val SAME_PRICE_PRODUCT = "same_price_product"
}

private object ApplicabilityRule {
  const val ALL_REQUIRED = "all_required"
  const val AT_LEAST_ONE_REQUIRED = "at_least_one_required"
}

private data class OrderItemInput(
  val id: String,
  val slug: String?,
  val quantity: Int,
  val originalPrice: Double,
  val promotionValue: Double?,
  val promotionDiscount: Double?,
  val name: String?,
  val image: String?,
  val size: String?,
  val note: String?,
  val productSlug: String?,
  val variant: Any?,
  val allVariants: Any?,
  val promotion: Any?,
  val description: String?,
  val isLimit: Boolean?,
  val isGift: Boolean?,
) {
  fun toMap(): MutableMap<String, Any?> {
    val m = mutableMapOf<String, Any?>(
      "id" to id,
      "slug" to slug,
      "quantity" to quantity,
      "originalPrice" to originalPrice,
    )
    name?.let { m["name"] = it }
    image?.let { m["image"] = it }
    size?.let { m["size"] = it }
    note?.let { m["note"] = it }
    productSlug?.let { m["productSlug"] = it }
    variant?.let { m["variant"] = it }
    allVariants?.let { m["allVariants"] = it }
    promotion?.let { m["promotion"] = it }
    description?.let { m["description"] = it }
    isLimit?.let { m["isLimit"] = it }
    isGift?.let { m["isGift"] = it }
    return m
  }

  companion object {
    fun fromJson(obj: JSONObject): OrderItemInput {
      fun getDouble(key: String): Double? = if (obj.has(key)) obj.getDouble(key) else null
      fun getInt(key: String): Int = if (obj.has(key)) obj.getInt(key) else 0
      fun getString(key: String): String? = if (obj.has(key)) obj.optString(key).takeIf { it != "null" } else null
      return OrderItemInput(
        id = obj.optString("id", ""),
        slug = getString("slug"),
        quantity = getInt("quantity"),
        originalPrice = obj.optDouble("originalPrice", 0.0),
        promotionValue = getDouble("promotionValue"),
        promotionDiscount = getDouble("promotionDiscount"),
        name = getString("name"),
        image = getString("image"),
        size = getString("size"),
        note = getString("note"),
        productSlug = getString("productSlug"),
        variant = if (obj.has("variant")) obj.get("variant") else null,
        allVariants = if (obj.has("allVariants")) obj.get("allVariants") else null,
        promotion = if (obj.has("promotion")) obj.get("promotion") else null,
        description = getString("description"),
        isLimit = if (obj.has("isLimit")) obj.getBoolean("isLimit") else null,
        isGift = if (obj.has("isGift")) obj.getBoolean("isGift") else null,
      )
    }
  }
}

private data class VoucherInput(
  val type: String,
  val value: Double,
  val applicabilityRule: String,
  val minOrderValue: Double,
  val voucherProducts: List<Map<String, Any>>,
) {
  companion object {
    fun fromJson(obj: JSONObject?): VoucherInput? {
      if (obj == null) return null
      val vps = mutableListOf<Map<String, Any>>()
      if (obj.has("voucherProducts")) {
        val arr = obj.getJSONArray("voucherProducts")
        for (i in 0 until arr.length()) {
          val vp = arr.getJSONObject(i)
          if (vp.has("product")) {
            val prod = vp.getJSONObject("product")
            val slug = prod.optString("slug", "")
            if (slug.isNotEmpty()) {
              vps.add(mapOf("product" to mapOf("slug" to slug)))
            }
          }
        }
      }
      return VoucherInput(
        type = obj.optString("type", ""),
        value = obj.optDouble("value", 0.0),
        applicabilityRule = obj.optString("applicabilityRule", ""),
        minOrderValue = obj.optDouble("minOrderValue", 0.0),
        voucherProducts = vps,
      )
    }
  }
}

private fun getRequiredSlugsSet(voucher: VoucherInput): Set<String> {
  return voucher.voucherProducts.mapNotNull { vp ->
    (vp["product"] as? Map<*, *>)?.get("slug") as? String
  }.filter { it.isNotEmpty() }.toSet()
}

private fun checkMinOrderValue(orderItems: List<OrderItemInput>, voucher: VoucherInput): Boolean {
  if (voucher.minOrderValue <= 0) return true
  val subtotal = orderItems.sumOf { item ->
    val orig = item.originalPrice
    val promo = item.promotionDiscount ?: 0.0
    (orig - promo) * item.quantity
  }
  return subtotal >= voucher.minOrderValue
}

private fun isVoucherApplicable(orderItems: List<OrderItemInput>, voucher: VoucherInput): Boolean {
  if (orderItems.isEmpty()) return false
  val requiredSlugs = getRequiredSlugsSet(voucher)
  if (requiredSlugs.isEmpty()) {
    return checkMinOrderValue(orderItems, voucher)
  }
  fun itemSlug(item: OrderItemInput) = item.slug ?: item.productSlug ?: ""
  if (voucher.applicabilityRule == ApplicabilityRule.ALL_REQUIRED) {
    if (!orderItems.all { requiredSlugs.contains(itemSlug(it)) }) return false
  } else if (voucher.applicabilityRule == ApplicabilityRule.AT_LEAST_ONE_REQUIRED) {
    if (!orderItems.any { requiredSlugs.contains(itemSlug(it)) }) return false
  }
  return checkMinOrderValue(orderItems, voucher)
}

private fun calculateCartItemDisplay(
  orderItems: List<OrderItemInput>,
  voucher: VoucherInput?,
): List<MutableMap<String, Any?>> {
  if (orderItems.isEmpty()) return emptyList()

  val (isVoucherValid, rule, type, eligibleSlugs) = if (voucher != null) {
    Quadruple(
      isVoucherApplicable(orderItems, voucher),
      voucher.applicabilityRule,
      voucher.type,
      getRequiredSlugsSet(voucher),
    )
  } else {
    Quadruple(false, "", "", emptySet())
  }

  val result = mutableListOf<MutableMap<String, Any?>>()
  for (item in orderItems) {
    val original = item.originalPrice
    val promoDiscount = item.promotionDiscount
      ?: (original * ((item.promotionValue ?: 0.0) / 100)).toInt().toDouble()
    val priceAfterPromo = maxOf(0.0, original - promoDiscount)
    val slug = item.slug ?: item.productSlug ?: ""
    val isEligible = eligibleSlugs.contains(slug)

    val out = item.toMap()
    out["priceAfterPromotion"] = priceAfterPromo
    out["promotionDiscount"] = promoDiscount

    if (!isVoucherValid || voucher == null) {
      out["finalPrice"] = priceAfterPromo
      out["voucherDiscount"] = 0
      result.add(out)
      continue
    }

    when (rule) {
      ApplicabilityRule.ALL_REQUIRED -> {
        when {
          type == VoucherType.PERCENT_ORDER && isEligible -> {
            out["finalPrice"] = priceAfterPromo
            out["voucherDiscount"] = 0
          }
          type == VoucherType.FIXED_VALUE && isEligible -> {
            out["finalPrice"] = priceAfterPromo
            out["voucherDiscount"] = 0
          }
          type == VoucherType.SAME_PRICE_PRODUCT && isEligible -> {
            val vVal = voucher.value
            val newPrice = if (vVal <= 1) (original * (1 - vVal)).toInt().toDouble()
            else minOf(original, vVal)
            out["finalPrice"] = newPrice
            out["priceAfterPromotion"] = priceAfterPromo
            out["promotionDiscount"] = 0
            out["voucherDiscount"] = original - newPrice
          }
          else -> {
            out["finalPrice"] = priceAfterPromo
            out["voucherDiscount"] = 0
          }
        }
      }
      ApplicabilityRule.AT_LEAST_ONE_REQUIRED -> {
        when {
          !isEligible -> {
            out["finalPrice"] = priceAfterPromo
            out["voucherDiscount"] = 0
          }
          type == VoucherType.PERCENT_ORDER -> {
            val vDiscount = ((voucher.value / 100) * original).toInt().toDouble()
            out["finalPrice"] = maxOf(0.0, original - vDiscount)
            out["priceAfterPromotion"] = original
            out["promotionDiscount"] = 0
            out["voucherDiscount"] = vDiscount
          }
          type == VoucherType.FIXED_VALUE -> {
            val vDiscount = minOf(original, voucher.value)
            out["finalPrice"] = maxOf(0.0, original - vDiscount)
            out["priceAfterPromotion"] = original
            out["promotionDiscount"] = 0
            out["voucherDiscount"] = vDiscount
          }
          type == VoucherType.SAME_PRICE_PRODUCT -> {
            val vVal = voucher.value
            val newPrice = if (vVal <= 1) (original * (1 - vVal)).toInt().toDouble()
            else minOf(original, vVal)
            out["finalPrice"] = newPrice
            out["priceAfterPromotion"] = original
            out["promotionDiscount"] = 0
            out["voucherDiscount"] = original - newPrice
          }
          else -> {
            out["finalPrice"] = priceAfterPromo
            out["voucherDiscount"] = 0
          }
        }
      }
      else -> {
        out["finalPrice"] = priceAfterPromo
        out["voucherDiscount"] = 0
      }
    }
    result.add(out)
  }
  return result
}

private data class Quadruple<A, B, C, D>(val first: A, val second: B, val third: C, val fourth: D)

private fun sumBy(items: List<Map<String, Any?>>, fn: (Map<String, Any?>) -> Double): Double {
  return items.sumOf { fn(it) }
}

private fun getDouble(map: Map<String, Any?>, key: String): Double {
  return when (val v = map[key]) {
    is Number -> v.toDouble()
    else -> 0.0
  }
}

private fun getInt(map: Map<String, Any?>, key: String): Int {
  return when (val v = map[key]) {
    is Number -> v.toInt()
    else -> 0
  }
}

private fun getString(map: Map<String, Any?>, key: String): String {
  return (map[key] as? String) ?: ""
}

private fun calculateCartTotals(
  displayItems: List<Map<String, Any?>>,
  voucher: VoucherInput?,
): Map<String, Any> {
  val allowedSlugs = (voucher?.voucherProducts?.mapNotNull { vp ->
    (vp["product"] as? Map<*, *>)?.get("slug") as? String
  }?.filter { it.isNotEmpty() } ?: emptyList()).toSet()

  val subTotalBeforeDiscount = sumBy(displayItems) { item ->
    getDouble(item, "originalPrice") * getInt(item, "quantity")
  }

  val promotionDiscount = sumBy(displayItems) { item ->
    val slug = getString(item, "slug").ifEmpty { getString(item, "productSlug") }
    val shouldExclude = (voucher?.type == VoucherType.SAME_PRICE_PRODUCT ||
      ((voucher?.type == VoucherType.PERCENT_ORDER || voucher?.type == VoucherType.FIXED_VALUE) &&
        voucher?.applicabilityRule == ApplicabilityRule.AT_LEAST_ONE_REQUIRED)) &&
      allowedSlugs.contains(slug) &&
      getDouble(item, "voucherDiscount") > 0

    val promo = getDouble(item, "promotionDiscount")
    val discount = if (!shouldExclude && promo > 0) promo else 0.0
    discount * getInt(item, "quantity")
  }

  var voucherDiscount = 0.0
  if (voucher != null) {
    val rule = voucher.applicabilityRule
    val type = voucher.type
    when (type) {
      VoucherType.SAME_PRICE_PRODUCT -> {
        voucherDiscount = sumBy(displayItems) { item ->
          val slug = getString(item, "slug").ifEmpty { getString(item, "productSlug") }
          if (allowedSlugs.contains(slug)) {
            getDouble(item, "voucherDiscount") * getInt(item, "quantity")
          } else 0.0
        }
      }
      VoucherType.PERCENT_ORDER -> {
        if (rule == ApplicabilityRule.ALL_REQUIRED) {
          val totalAfterPromo = subTotalBeforeDiscount - promotionDiscount
          voucherDiscount = ((voucher.value / 100) * totalAfterPromo).toInt().toDouble()
        } else {
          voucherDiscount = sumBy(displayItems) { item ->
            getDouble(item, "voucherDiscount") * getInt(item, "quantity")
          }
        }
      }
      VoucherType.FIXED_VALUE -> {
        if (rule == ApplicabilityRule.ALL_REQUIRED) {
          val totalAfterPromo = subTotalBeforeDiscount - promotionDiscount
          voucherDiscount = minOf(voucher.value, totalAfterPromo)
        } else {
          voucherDiscount = sumBy(displayItems) { item ->
            getDouble(item, "voucherDiscount") * getInt(item, "quantity")
          }
        }
      }
    }
  }

  val finalTotal = subTotalBeforeDiscount - promotionDiscount - voucherDiscount

  return mapOf(
    "subTotalBeforeDiscount" to subTotalBeforeDiscount,
    "promotionDiscount" to promotionDiscount,
    "voucherDiscount" to voucherDiscount,
    "finalTotal" to finalTotal,
  )
}

private fun mapToJsonArray(items: List<Map<String, Any?>>): JSONArray {
  val arr = JSONArray()
  for (item in items) {
    arr.put(mapToJsonObject(item))
  }
  return arr
}

private fun mapToJsonObject(m: Map<String, Any?>): JSONObject {
  val obj = JSONObject()
  for ((k, v) in m) {
    when (v) {
      null -> obj.put(k, JSONObject.NULL)
      is Number -> obj.put(k, v)
      is Boolean -> obj.put(k, v)
      is String -> obj.put(k, v)
      is Map<*, *> -> obj.put(k, mapToJsonObject(v as Map<String, Any?>))
      is List<*> -> obj.put(k, listToJsonArray(v))
      is org.json.JSONObject -> obj.put(k, v)
      is org.json.JSONArray -> obj.put(k, v)
      else -> obj.put(k, v?.toString() ?: "")
    }
  }
  return obj
}

private fun listToJsonArray(list: List<*>): JSONArray {
  val arr = JSONArray()
  for (v in list) {
    when (v) {
      null -> arr.put(JSONObject.NULL)
      is Number -> arr.put(v)
      is Boolean -> arr.put(v)
      is String -> arr.put(v)
      is Map<*, *> -> arr.put(mapToJsonObject(v as Map<String, Any?>))
      is List<*> -> arr.put(listToJsonArray(v))
      is org.json.JSONObject -> arr.put(v)
      is org.json.JSONArray -> arr.put(v)
      else -> arr.put(v?.toString() ?: "")
    }
  }
  return arr
}

private val emptyResultJson = """{"displayItems":[],"totals":{"subTotalBeforeDiscount":0,"promotionDiscount":0,"voucherDiscount":0,"finalTotal":0}}"""

fun calculateRawSubTotal(orderItemsJson: String): Double {
  return try {
    val itemsArray = JSONArray(orderItemsJson)
    var sum = 0.0
    for (i in 0 until itemsArray.length()) {
      val item = OrderItemInput.fromJson(itemsArray.getJSONObject(i))
      sum += item.originalPrice * item.quantity
    }
    sum
  } catch (_: Exception) {
    0.0
  }
}

private val viNumberFormat =
  java.text.DecimalFormat("#,##0.##", java.text.DecimalFormatSymbols(java.util.Locale("vi", "VN"))).apply {
    minimumFractionDigits = 0
    maximumFractionDigits = 0
  }

fun formatCurrencyNative(value: Double, currency: String): String {
  val safeValue = if (value < 0) 0.0 else value
  return "${viNumberFormat.format(safeValue)} $currency"
}

fun computeProductDetailTotalNative(
  price: Double,
  quantity: Int,
  toppingExtra: Double,
  promotionPercent: Double,
): Double {
  if (price < 0) return 0.0
  val afterDiscount = price * (1 - promotionPercent / 100)
  return (afterDiscount + toppingExtra) * quantity
}

fun calculateDisplayItemsNative(orderItemsJson: String, voucherJson: String?): String {
  return try {
    val itemsArray = JSONArray(orderItemsJson)
    val orderItems = List(itemsArray.length()) { i ->
      OrderItemInput.fromJson(itemsArray.getJSONObject(i))
    }

    val voucher = when {
      voucherJson.isNullOrEmpty() -> null
      else -> try {
        VoucherInput.fromJson(JSONObject(voucherJson))
      } catch (_: Exception) {
        null
      }
    }

    val displayItems = calculateCartItemDisplay(orderItems, voucher)
    val totals = calculateCartTotals(displayItems, voucher)

    val result = JSONObject().apply {
      put("displayItems", mapToJsonArray(displayItems))
      put("totals", mapToJsonObject(totals))
    }
    result.toString()
  } catch (_: Exception) {
    emptyResultJson
  }
}

File: modules/cart-price-calc/ios/CartPriceCalculator.swift
/**
 * Cart Price Calculator — Native port của utils/cart.ts
 * Logic voucher: ALL_REQUIRED, AT_LEAST_ONE_REQUIRED × PERCENT_ORDER, FIXED_VALUE, SAME_PRICE_PRODUCT
 */
import Foundation

private enum VoucherType {
  static let percentOrder = "percent_order"
  static let fixedValue = "fixed_value"
  static let samePriceProduct = "same_price_product"
}

private enum ApplicabilityRule {
  static let allRequired = "all_required"
  static let atLeastOneRequired = "at_least_one_required"
}

struct OrderItemInput {
  var id: String
  var slug: String?
  var quantity: Int
  var originalPrice: Double
  var promotionValue: Double?
  var promotionDiscount: Double?
  var name: String?
  var image: String?
  var size: String?
  var note: String?
  var productSlug: String?
  var variant: [String: Any]?
  var allVariants: [[String: Any]]?
  var promotion: [String: Any]?
  var promotionValueNum: Double?
  var description: String?
  var isLimit: Bool?
  var isGift: Bool?

  init(from dict: [String: Any]) {
    id = dict["id"] as? String ?? ""
    slug = dict["slug"] as? String
    quantity = (dict["quantity"] as? NSNumber)?.intValue ?? 0
    originalPrice = (dict["originalPrice"] as? NSNumber)?.doubleValue ?? 0
    promotionValue = (dict["promotionValue"] as? NSNumber)?.doubleValue
    promotionDiscount = (dict["promotionDiscount"] as? NSNumber)?.doubleValue
    name = dict["name"] as? String
    image = dict["image"] as? String
    size = dict["size"] as? String
    note = dict["note"] as? String
    productSlug = dict["productSlug"] as? String
    variant = dict["variant"] as? [String: Any]
    allVariants = dict["allVariants"] as? [[String: Any]]
    promotion = dict["promotion"] as? [String: Any]
    promotionValueNum = (dict["promotionValue"] as? NSNumber)?.doubleValue
    description = dict["description"] as? String
    isLimit = dict["isLimit"] as? Bool
    isGift = dict["isGift"] as? Bool
  }

  func toDict() -> [String: Any] {
    var d: [String: Any] = [
      "id": id,
      "slug": slug as Any,
      "quantity": quantity,
      "originalPrice": originalPrice,
    ]
    if let v = name { d["name"] = v }
    if let v = image { d["image"] = v }
    if let v = size { d["size"] = v }
    if let v = note { d["note"] = v }
    if let v = productSlug { d["productSlug"] = v }
    if let v = variant { d["variant"] = v }
    if let v = allVariants { d["allVariants"] = v }
    if let v = promotion { d["promotion"] = v }
    if let v = description { d["description"] = v }
    if let v = isLimit { d["isLimit"] = v }
    if let v = isGift { d["isGift"] = v }
    return d
  }
}

struct VoucherInput {
  var type: String
  var value: Double
  var applicabilityRule: String
  var minOrderValue: Double
  var voucherProducts: [[String: Any]]

  init?(from dict: [String: Any]?) {
    guard let d = dict else { return nil }
    type = d["type"] as? String ?? ""
    value = (d["value"] as? NSNumber)?.doubleValue ?? 0
    applicabilityRule = d["applicabilityRule"] as? String ?? ""
    minOrderValue = (d["minOrderValue"] as? NSNumber)?.doubleValue ?? 0
    let vps = d["voucherProducts"] as? [[String: Any]] ?? []
    voucherProducts = vps.compactMap { vp -> [String: Any]? in
      guard let prod = vp["product"] as? [String: Any],
            let slug = prod["slug"] as? String, !slug.isEmpty else { return nil }
      return ["product": ["slug": slug]]
    }
  }
}

// MARK: - Helpers

private func getRequiredSlugsSet(voucher: VoucherInput) -> Set<String> {
  let slugs = voucher.voucherProducts.compactMap { vp -> String? in
    guard let prod = vp["product"] as? [String: Any],
          let slug = prod["slug"] as? String, !slug.isEmpty else { return nil }
    return slug
  }
  return Set(slugs)
}

private func checkMinOrderValue(orderItems: [OrderItemInput], voucher: VoucherInput) -> Bool {
  if voucher.minOrderValue <= 0 { return true }
  var subtotal: Double = 0
  for item in orderItems {
    let orig = item.originalPrice
    let promo = item.promotionDiscount ?? 0
    subtotal += (orig - promo) * Double(item.quantity)
  }
  return subtotal >= voucher.minOrderValue
}

private func isVoucherApplicable(orderItems: [OrderItemInput], voucher: VoucherInput) -> Bool {
  if orderItems.isEmpty { return false }
  let requiredSlugs = getRequiredSlugsSet(voucher: voucher)
  if requiredSlugs.isEmpty {
    return checkMinOrderValue(orderItems: orderItems, voucher: voucher)
  }
  let itemSlug: (OrderItemInput) -> String = { $0.slug ?? $0.productSlug ?? "" }
  if voucher.applicabilityRule == ApplicabilityRule.allRequired {
    let allIn = orderItems.allSatisfy { requiredSlugs.contains(itemSlug($0)) }
    if !allIn { return false }
  } else if voucher.applicabilityRule == ApplicabilityRule.atLeastOneRequired {
    let hasOne = orderItems.contains { requiredSlugs.contains(itemSlug($0)) }
    if !hasOne { return false }
  }
  return checkMinOrderValue(orderItems: orderItems, voucher: voucher)
}

// MARK: - calculateCartItemDisplay

private func calculateCartItemDisplay(
  orderItems: [OrderItemInput],
  voucher: VoucherInput?
) -> [[String: Any]] {
  guard !orderItems.isEmpty else { return [] }

  let isVoucherValid: Bool
  let rule: String
  let type: String
  let eligibleSlugs: Set<String>

  if let v = voucher {
    isVoucherValid = isVoucherApplicable(orderItems: orderItems, voucher: v)
    rule = v.applicabilityRule
    type = v.type
    eligibleSlugs = getRequiredSlugsSet(voucher: v)
  } else {
    isVoucherValid = false
    rule = ""
    type = ""
    eligibleSlugs = []
  }

  var result: [[String: Any]] = []
  for item in orderItems {
    let original = item.originalPrice
    let promoDiscount = item.promotionDiscount ?? Double(Int(original * ((item.promotionValue ?? 0) / 100)))
    let priceAfterPromo = max(0, original - promoDiscount)
    let slug = item.slug ?? item.productSlug ?? ""
    let isEligible = eligibleSlugs.contains(slug)

    var out = item.toDict()
    out["priceAfterPromotion"] = priceAfterPromo
    out["promotionDiscount"] = promoDiscount

    if !isVoucherValid || voucher == nil {
      out["finalPrice"] = priceAfterPromo
      out["voucherDiscount"] = 0
      result.append(out)
      continue
    }

    if rule == ApplicabilityRule.allRequired {
      if type == VoucherType.percentOrder && isEligible {
        out["finalPrice"] = priceAfterPromo
        out["voucherDiscount"] = 0
      } else if type == VoucherType.fixedValue && isEligible {
        out["finalPrice"] = priceAfterPromo
        out["voucherDiscount"] = 0
      } else if type == VoucherType.samePriceProduct && isEligible {
        let vVal = voucher!.value
        let newPrice = vVal <= 1
          ? Double(Int(original * (1 - vVal)))
          : min(original, vVal)
        out["finalPrice"] = newPrice
        out["priceAfterPromotion"] = priceAfterPromo
        out["promotionDiscount"] = 0
        out["voucherDiscount"] = original - newPrice
      } else {
        out["finalPrice"] = priceAfterPromo
        out["voucherDiscount"] = 0
      }
    } else if rule == ApplicabilityRule.atLeastOneRequired {
      if !isEligible {
        out["finalPrice"] = priceAfterPromo
        out["voucherDiscount"] = 0
      } else if type == VoucherType.percentOrder {
        let vDiscount = Double(Int(((voucher!.value) / 100) * original))
        out["finalPrice"] = max(0, original - vDiscount)
        out["priceAfterPromotion"] = original
        out["promotionDiscount"] = 0
        out["voucherDiscount"] = vDiscount
      } else if type == VoucherType.fixedValue {
        let vDiscount = min(original, voucher!.value)
        out["finalPrice"] = max(0, original - vDiscount)
        out["priceAfterPromotion"] = original
        out["promotionDiscount"] = 0
        out["voucherDiscount"] = vDiscount
      } else if type == VoucherType.samePriceProduct {
        let vVal = voucher!.value
        let newPrice = vVal <= 1
          ? Double(Int(original * (1 - vVal)))
          : min(original, vVal)
        out["finalPrice"] = newPrice
        out["priceAfterPromotion"] = original
        out["promotionDiscount"] = 0
        out["voucherDiscount"] = original - newPrice
      } else {
        out["finalPrice"] = priceAfterPromo
        out["voucherDiscount"] = 0
      }
    } else {
      out["finalPrice"] = priceAfterPromo
      out["voucherDiscount"] = 0
    }
    result.append(out)
  }
  return result
}

// MARK: - calculateCartTotals

private func sumBy(_ items: [[String: Any]], fn: ([String: Any]) -> Double) -> Double {
  var sum: Double = 0
  for item in items {
    sum += fn(item)
  }
  return sum
}

private func calculateCartTotals(
  displayItems: [[String: Any]],
  voucher: VoucherInput?
) -> [String: Any] {
  let allowedSlugs = Set(
    (voucher?.voucherProducts ?? []).compactMap { vp -> String? in
      (vp["product"] as? [String: Any])?["slug"] as? String
    }.filter { !$0.isEmpty }
  )

  let subTotalBeforeDiscount = sumBy(displayItems) { item in
    let orig = (item["originalPrice"] as? NSNumber)?.doubleValue ?? 0
    let qty = (item["quantity"] as? NSNumber)?.intValue ?? 0
    return orig * Double(qty)
  }

  let promotionDiscount = sumBy(displayItems) { item in
    let slug = item["slug"] as? String ?? item["productSlug"] as? String ?? ""
    let shouldExclude = (voucher?.type == VoucherType.samePriceProduct ||
      ((voucher?.type == VoucherType.percentOrder || voucher?.type == VoucherType.fixedValue) &&
        voucher?.applicabilityRule == ApplicabilityRule.atLeastOneRequired)) &&
      allowedSlugs.contains(slug) &&
      ((item["voucherDiscount"] as? NSNumber)?.doubleValue ?? 0) > 0

    let promo = (item["promotionDiscount"] as? NSNumber)?.doubleValue ?? 0
    let discount = !shouldExclude && promo > 0 ? promo : 0
    let qty = (item["quantity"] as? NSNumber)?.intValue ?? 0
    return discount * Double(qty)
  }

  var voucherDiscount: Double = 0
  if let v = voucher {
    let rule = v.applicabilityRule
    let type = v.type
    if type == VoucherType.samePriceProduct {
      voucherDiscount = sumBy(displayItems) { item in
        let slug = item["slug"] as? String ?? item["productSlug"] as? String ?? ""
        if allowedSlugs.contains(slug) {
          let vd = (item["voucherDiscount"] as? NSNumber)?.doubleValue ?? 0
          let qty = (item["quantity"] as? NSNumber)?.intValue ?? 0
          return vd * Double(qty)
        }
        return 0
      }
    } else if type == VoucherType.percentOrder {
      if rule == ApplicabilityRule.allRequired {
        let totalAfterPromo = subTotalBeforeDiscount - promotionDiscount
        voucherDiscount = Double(Int((v.value / 100) * totalAfterPromo))
      } else {
        voucherDiscount = sumBy(displayItems) { item in
          let vd = (item["voucherDiscount"] as? NSNumber)?.doubleValue ?? 0
          let qty = (item["quantity"] as? NSNumber)?.intValue ?? 0
          return vd * Double(qty)
        }
      }
    } else if type == VoucherType.fixedValue {
      if rule == ApplicabilityRule.allRequired {
        let totalAfterPromo = subTotalBeforeDiscount - promotionDiscount
        voucherDiscount = min(v.value, totalAfterPromo)
      } else {
        voucherDiscount = sumBy(displayItems) { item in
          let vd = (item["voucherDiscount"] as? NSNumber)?.doubleValue ?? 0
          let qty = (item["quantity"] as? NSNumber)?.intValue ?? 0
          return vd * Double(qty)
        }
      }
    }
  }

  let finalTotal = subTotalBeforeDiscount - promotionDiscount - voucherDiscount

  return [
    "subTotalBeforeDiscount": subTotalBeforeDiscount,
    "promotionDiscount": promotionDiscount,
    "voucherDiscount": voucherDiscount,
    "finalTotal": finalTotal,
  ]
}

func calculateRawSubTotal(orderItemsJson: String) -> Double {
  guard let itemsData = orderItemsJson.data(using: .utf8),
        let itemsArray = try? JSONSerialization.jsonObject(with: itemsData) as? [[String: Any]] else {
    return 0
  }
  return itemsArray.reduce(0) { sum, dict in
    let item = OrderItemInput(from: dict)
    return sum + item.originalPrice * Double(item.quantity)
  }
}

private let viNumberFormatter: NumberFormatter = {
  let f = NumberFormatter()
  f.locale = Locale(identifier: "vi_VN")
  f.numberStyle = .decimal
  f.minimumFractionDigits = 0
  f.maximumFractionDigits = 0
  return f
}()

func formatCurrencyNative(value: Double, currency: String) -> String {
  let safeValue = value < 0 ? 0 : value
  let formatted = viNumberFormatter.string(from: NSNumber(value: safeValue)) ?? "\(Int(safeValue))"
  return "\(formatted) \(currency)"
}

func computeProductDetailTotalNative(
  price: Double,
  quantity: Int,
  toppingExtra: Double,
  promotionPercent: Double
) -> Double {
  guard price >= 0 else { return 0 }
  let afterDiscount = price * (1 - promotionPercent / 100)
  return (afterDiscount + toppingExtra) * Double(quantity)
}

// MARK: - Public

func calculateDisplayItemsNative(
  orderItemsJson: String,
  voucherJson: String?
) -> String {
  guard let itemsData = orderItemsJson.data(using: .utf8),
        let itemsArray = try? JSONSerialization.jsonObject(with: itemsData) as? [[String: Any]] else {
    return emptyResultJson
  }

  let orderItems = itemsArray.map { OrderItemInput(from: $0) }

  var voucher: VoucherInput?
  if let vStr = voucherJson, !vStr.isEmpty,
     let vData = vStr.data(using: .utf8),
     let vDict = try? JSONSerialization.jsonObject(with: vData) as? [String: Any] {
    voucher = VoucherInput(from: vDict)
  }

  let displayItems = calculateCartItemDisplay(orderItems: orderItems, voucher: voucher)
  let totals = calculateCartTotals(displayItems: displayItems, voucher: voucher)

  let result: [String: Any] = [
    "displayItems": displayItems,
    "totals": totals,
  ]

  guard let data = try? JSONSerialization.data(withJSONObject: result),
        let str = String(data: data, encoding: .utf8) else {
    return emptyResultJson
  }
  return str
}

private let emptyResultJson = "{\"displayItems\":[],\"totals\":{\"subTotalBeforeDiscount\":0,\"promotionDiscount\":0,\"voucherDiscount\":0,\"finalTotal\":0}}"
